File: LeetCode75_LevelTwo/Binary_Search_Tree_Iterator/Binary_Search_Tree_Iterator.cc
// 173. Binary Search Tree Iterator C++
#include <vector>

 // Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class BSTIterator {
    std::vector<int> v;
    int index;
public:
    BSTIterator(TreeNode* root) {
        helper(root);
        index = -1;
    }
    void helper(TreeNode* root) {
        if (!root) return;
        if (root->left) helper(root->left);
        v.emplace_back(root->val);
        if (root->right) helper(root->right);
    }
    
    int next() {
        return v[++index];
    }
    
    bool hasNext() {
        return index + 1 < v.size();
    }
};

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator* obj = new BSTIterator(root);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */

File: LeetCode75_LevelTwo/Binary_Search_Tree_Iterator/Binary_Search_Tree_Iterator.py
# 173. Binary Search Tree Iterator Python

# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class BSTIterator(object):

    def __init__(self, root):
        """
        :type root: TreeNode
        """
        self.nodes_sorted = []
        self.index = -1
        self._inorder(root)
    def _inorder(self, root) :
        if not root: return
        self._inorder(root.left)
        self.nodes_sorted.append(root.val)
        self._inorder(root.right)

    def next(self):
        """
        :rtype: int
        """
        self.index += 1
        return self.nodes_sorted[self.index]

    def hasNext(self):
        """
        :rtype: bool
        """
        return self.index + 1 < len(self.nodes_sorted)
        


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()